### PR TITLE
OSSM-3987 Increase test coverage cluster wide 

### PR DESCRIPTION
**Pull request by fjglira**
Jira: https://issues.redhat.com/browse/OSSM-3987

Adding new test case to increase test coverage:

- Check Rolebindings: https://issues.redhat.com/browse/OSSM-3468

### CODE REVIEW

**Comment by fjglira** [comment]
Work done and ready to review

**Comment by fjglira** [comment]
Adding change in sub test case: cluster wide works with profiles. Going to use this profile: `gateway-controller` to meet this Jira: https://issues.redhat.com/browse/OSSM-3562

File: pkg/tests/ossm/operator/clusterwide_mode_test.go [hunk 1]
```diff
@@ -70,6 +70,29 @@ func TestClusterWideMode(t *testing.T) {
...
+				oc.Get(t, "member-0", "rolebindings", "",
+					assert.OutputContains("system:deployers",
+						"The Rolebings contains system:deployers role",
+						"The Rolebings does not contains system:deployers role"),
+					assert.OutputContains("system:image-builders",
+						"The Rolebings contains system:image-builders role",
+						"The Rolebings does not contains system:image-builders role"),
+					assert.OutputContains("system:image-pullers",
+						"The Rolebings contains system:image-pullers role",
+						"The Rolebings does not contains system:image-pullers role"),
```
Comment (luksa) [inline pkg/tests/ossm/operator/clusterwide_mode_test.go:86]: Why check these rolebindings? They're not related to OSSM, so I think we should ignore them.

File: pkg/tests/ossm/operator/clusterwide_mode_test.go [hunk 2]
```diff
@@ -70,6 +70,29 @@ func TestClusterWideMode(t *testing.T) {
...
+					assert.OutputDoesNotContain("istiod-clusterrole-basic-istio-system",
+						"The Rolebings does not contains istiod-clusterrole-basic-istio-system role",
+						"The Rolebings contains istiod-clusterrole-basic-istio-system role"),
```
Comment (luksa) [inline pkg/tests/ossm/operator/clusterwide_mode_test.go:89]:
```suggestion
						"The istiod-clusterrole-basic-istio-system RoleBinding does not exist",
						"The istiod-clusterrole-basic-istio-system RoleBinding exists"),
```

File: pkg/tests/ossm/operator/clusterwide_mode_test.go [hunk 3]
```diff
@@ -70,6 +70,29 @@ func TestClusterWideMode(t *testing.T) {
...
+					assert.OutputDoesNotContain("istiod-gateway-controller-basic-istio-system",
+						"The Rolebings does not contains istiod-gateway-controller-basic-istio-system role",
+						"The Rolebings contains istiod-gateway-controller-basic-istio-system role"))
```
Comment (luksa) [inline pkg/tests/ossm/operator/clusterwide_mode_test.go:92]: Same here

File: pkg/tests/ossm/operator/clusterwide_mode_test.go [hunk 4]
```diff
@@ -70,6 +70,29 @@ func TestClusterWideMode(t *testing.T) {
...
+			})
+		})
+
 		t.NewSubTest("customize SMMR").Run(func(t test.TestHelper) {
```
Comment (luksa) [inline pkg/tests/ossm/operator/clusterwide_mode_test.go:96]: Shouldn't this be multiple subtests, since many different things are tested here? It's also hard to follow what this subtest does, because it's so long.

File: pkg/tests/ossm/operator/clusterwide_mode_test.go [hunk 5]
```diff
@@ -104,6 +127,19 @@ func TestClusterWideMode(t *testing.T) {
...
+			oc.ApplyString(t, meshNamespace, smmrMultipleSelectors)
+			oc.WaitSMMRReady(t, meshNamespace)
+
+			t.LogStep("Check whether the SMMR shows only namespaces as members: member-0")
+			retry.UntilSuccess(t, func(t test.TestHelper) {
+				shell.Execute(t,
+					fmt.Sprintf("oc -n %s get smmr default", meshNamespace),
+					require.OutputContains("1/1",
+						"one namespace are in members",
+						"expected SMMR to show 1 member namespace, but that wasn't the case"))
```
Comment (luksa) [inline pkg/tests/ossm/operator/clusterwide_mode_test.go:140]: Why not check the actual namespace name instead of just the namespace count? The operator could add a completely different namespace here and the test would still pass.

Comment (luksa) [reply pkg/tests/ossm/operator/clusterwide_mode_test.go:140]: BTW, why do we create 50 member namespaces? Seems unnecessary. 5 would be more than enough.

File: pkg/tests/ossm/operator/clusterwide_mode_test.go [hunk 6]
```diff
@@ -172,33 +208,36 @@ func TestClusterWideMode(t *testing.T) {
...
-		// 			assert.OutputContains("default",
-		// 				"The SMMR was created immediately after the SMCP was created",
-		// 				"The SMMR resource was not created"))
-		// 	})
-		// })
+		t.NewSubTest("cluster wide works with profiles").Run(func(t test.TestHelper) {
+			t.Log("Check whether the cluster wide feature works with profiles")
+
+			t.LogStep("Delete SMCP and SMMR")
+			oc.RecreateNamespace(t, meshNamespace)
```
Comment (luksa) [inline pkg/tests/ossm/operator/clusterwide_mode_test.go:215]: Note: deleting and recreating the namespace takes a long time because of Kiali. Whenever possible, delete the individual resources instead of the entire namespace.

File: pkg/tests/ossm/operator/clusterwide_mode_test.go (hunk 4 above)
Comment (fjglira) [reply pkg/tests/ossm/operator/clusterwide_mode_test.go:96]: The only thing to not break in more sub test cases at the beginning is the dependency between the steps, but sure I can split in more subtest cases

File: pkg/tests/ossm/operator/clusterwide_mode_test.go (hunk 5 above)
Comment (fjglira) [reply pkg/tests/ossm/operator/clusterwide_mode_test.go]: Improved this

File: pkg/tests/ossm/operator/clusterwide_mode_test.go (hunk 6 above)
Comment (fjglira) [reply pkg/tests/ossm/operator/clusterwide_mode_test.go]: Improved this

**Comment by fjglira** [comment]
Imrpove execution time by non recreating the entire namespace

```
--- PASS: TestClusterWideMode (162.04s)
    --- PASS: TestClusterWideMode/SMMR_auto-creation (0.52s)
    --- PASS: TestClusterWideMode/default_namespace_selector (3.44s)
    --- PASS: TestClusterWideMode/RoleBindings_verification (0.52s)
    --- PASS: TestClusterWideMode/customize_SMMR (3.19s)
    --- PASS: TestClusterWideMode/verify_memberselector_operator_IN (3.25s)
    --- PASS: TestClusterWideMode/verify_multiple_memberselector (2.74s)
    --- PASS: TestClusterWideMode/verify_memberselector_operator_NOTIN (5.63s)
    --- PASS: TestClusterWideMode/verify_sidecar_injection (6.93s)
    --- PASS: TestClusterWideMode/cluster-scoped_watches_in_istiod (7.88s)
    --- PASS: TestClusterWideMode/cluster_wide_works_with_profiles (10.18s)
PASS
ok      github.com/maistra/maistra-test-tool/pkg/tests/ossm/operator    164.104s
```

**Comment by fjglira** [comment]
@marko all the changes done

File: pkg/tests/ossm/operator/clusterwide_mode_test.go [hunk 7]
```diff
@@ -77,57 +101,70 @@ func TestClusterWideMode(t *testing.T) {
...
+			membersList := []string{"member-1", "member-2", "member-3", "member-4"}
+			notInMembersList := []string{"member-0"}
 			retry.UntilSuccess(t, func(t test.TestHelper) {
-				shell.Execute(t,
-					fmt.Sprintf("oc -n %s get smmr default", meshNamespace),
-					require.OutputContains("55/55",
-						"All the namespaces are in members except member-0",
-						"expected SMMR to show 55 member namespace, but that wasn't the case"))
+				verifyMembersInSMMR(t, meshNamespace, membersList, true)
+				verifyMembersInSMMR(t, meshNamespace, notInMembersList, false)
```
Comment (luksa) [inline pkg/tests/ossm/operator/clusterwide_mode_test.go:157]: The function name is misleading when you pass in `false`. Please create two functions (e.g. `assertMembers`, `assertNonMembers` or something similar).

**Comment by luksa** [comment]
What's the purpose of the `clusterScoped` file? Why is it called that and not `clusterWide`? Also, why doesn't it have the `.yaml` extension?

File: pkg/tests/ossm/operator/clusterwide_mode_test.go (hunk 7 above)
Comment (fjglira) [reply pkg/tests/ossm/operator/clusterwide_mode_test.go:157]: Done

**Comment by fjglira** [comment]
> What's the purpose of the `clusterScoped` file? Why is it called that and not `clusterWide`? Also, why doesn't it have the `.yaml` extension?

Deleted, was used before but when I added to work with gateway profile was unused.

**Comment by fjglira** [comment]
Tested
```
--- PASS: TestClusterWideMode (120.15s)
    --- PASS: TestClusterWideMode/SMMR_auto-creation (0.61s)
    --- PASS: TestClusterWideMode/default_namespace_selector (4.20s)
    --- PASS: TestClusterWideMode/RoleBindings_verification (0.63s)
    --- PASS: TestClusterWideMode/customize_SMMR (3.32s)
    --- PASS: TestClusterWideMode/verify_memberselector_operator_IN (3.38s)
    --- PASS: TestClusterWideMode/verify_multiple_memberselector (3.29s)
    --- PASS: TestClusterWideMode/verify_memberselector_operator_NOTIN (6.20s)
    --- PASS: TestClusterWideMode/verify_sidecar_injection (14.65s)
    --- PASS: TestClusterWideMode/cluster-scoped_watches_in_istiod (6.88s)
    --- PASS: TestClusterWideMode/cluster_wide_works_with_profiles (11.96s)
```

File: pkg/tests/ossm/operator/clusterwide_mode_test.go [hunk 8]
```diff
@@ -172,36 +198,80 @@ func TestClusterWideMode(t *testing.T) {
...
+		found := false
+		for _, m := range members {
+			if member == m {
+				found = true
+				break
+			}
+		}
+		if found != shouldExist {
+			if shouldExist {
+				t.Fatalf("The member '%s' is missing from the members list.\n", member)
```
Comment (luksa) [inline pkg/tests/ossm/operator/clusterwide_mode_test.go:267]:
```suggestion
				t.Fatalf("The member '%s' is missing from the members list.", member)
```

File: pkg/tests/ossm/operator/clusterwide_mode_test.go [hunk 9]
```diff
@@ -172,36 +198,80 @@ func TestClusterWideMode(t *testing.T) {
...
+			} else {
+				t.Fatalf("The member '%s' exists in the members list.\n", member)
```
Comment (luksa) [inline pkg/tests/ossm/operator/clusterwide_mode_test.go:269]:
```suggestion
				t.Fatalf("The member '%s' exists in the members list.", member)
```

File: pkg/tests/ossm/operator/clusterwide_mode_test.go [hunk 10]
```diff
@@ -172,36 +198,80 @@ func TestClusterWideMode(t *testing.T) {
...
+
+func assertNonMembers(t test.TestHelper, meshNamespace string, membersList []string) {
+	retry.UntilSuccess(t, func(t test.TestHelper) {
+		verifyMembersInSMMR(t, meshNamespace, membersList, false)
+	})
+}
+
+// verifyMembersInSMMR verifies whether the SMMR has or not have the members provided in the members list
+func verifyMembersInSMMR(t test.TestHelper, meshNamespace string, membersList []string, shouldExist bool) {
+	smcpYaml := oc.GetYaml(t, meshNamespace, "smmr", "default")
```
Comment (luksa) [inline pkg/tests/ossm/operator/clusterwide_mode_test.go:249]: Misleading name. It's a SMMR, not SMCP.

Comment (fjglira) [reply pkg/tests/ossm/operator/clusterwide_mode_test.go]: Done

File: pkg/tests/ossm/operator/clusterwide_mode_test.go [hunk 11]
```diff
@@ -56,18 +64,33 @@ func TestClusterWideMode(t *testing.T) {
...
+			assertMembers(t, meshNamespace, membersList)
+
+		})
+
+		t.NewSubTest("RoleBindings verification").Run(func(t test.TestHelper) {
+			t.Log("Related to OSSM-3468")
+			t.LogStep("Check that Rolebindings are not created in the member namespaces")
+			retry.UntilSuccess(t, func(t test.TestHelper) {
+				oc.Get(t, "member-0", "rolebindings", "",
+					assert.OutputContains("prometheus-istio-system",
```
Comment (jewertow) [inline pkg/tests/ossm/operator/clusterwide_mode_test.go:84]: This role binding was removed, but then I [reverted](https://github.com/maistra/istio-operator/pull/1194) that change, so now this test should fail. Actually, since we will no longer use our prometheus, we can revert my commit to remove that role binding back, but it's too late and it has to be postponed to 2.4.1.
Anyway, this assertion should fail and you will have to remove it for now.

**Review by jewertow — CHANGES_REQUESTED** [review]

**Review by jewertow — APPROVED** [review]

**Comment by fjglira** [comment]
Test passed with latest changes
```
--- PASS: TestClusterWideMode (136.59s)
    --- PASS: TestClusterWideMode/SMMR_auto-creation (0.62s)
    --- PASS: TestClusterWideMode/default_namespace_selector (4.21s)
    --- PASS: TestClusterWideMode/RoleBindings_verification (0.63s)
    --- PASS: TestClusterWideMode/customize_SMMR (3.44s)
    --- PASS: TestClusterWideMode/verify_memberselector_operator_IN (3.43s)
    --- PASS: TestClusterWideMode/verify_multiple_memberselector (4.71s)
    --- PASS: TestClusterWideMode/verify_memberselector_operator_NOTIN (16.19s)
    --- PASS: TestClusterWideMode/verify_sidecar_injection (15.56s)
    --- PASS: TestClusterWideMode/cluster-scoped_watches_in_istiod (10.28s)
    --- PASS: TestClusterWideMode/cluster_wide_works_with_profiles (12.91s)
```
